pkg/planner/core/rule_outer_to_inner_join.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2024 PingCAP, Inc.
+
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+
+	"github.com/pingcap/tidb/pkg/expression"
+	"github.com/pingcap/tidb/pkg/parser/ast"
+	"github.com/pingcap/tidb/pkg/planner/core/base"
+	"github.com/pingcap/tidb/pkg/planner/util"
+	"github.com/pingcap/tidb/pkg/planner/util/optimizetrace"
+)
+
+func mergeOnClausePredicates(p *LogicalJoin, predicates []expression.Expression) []expression.Expression {
+	combinedCond := make([]expression.Expression, 0,
+		len(p.LeftConditions)+len(p.RightConditions)+
+			len(p.EqualConditions)+len(p.OtherConditions)+
+			len(predicates))
+	combinedCond = append(combinedCond, p.LeftConditions...)
+	combinedCond = append(combinedCond, p.RightConditions...)
+	combinedCond = append(combinedCond, expression.ScalarFuncs2Exprs(p.EqualConditions)...)
+	combinedCond = append(combinedCond, p.OtherConditions...)
+	combinedCond = append(combinedCond, predicates...)
+	return combinedCond
+}
+
+// convertOuterToInnerJoin converts outer to inner joins if the unmtaching rows are filtered.
```

Comment:
Could you please add more detail comments in here such as the query example, main optimization process (Preorder traversal or subsequent traversal?)?

Reply:
OK. Added more comments.